File: Turbolinks/Turbolinks/TLWebView.swift
import WebKit

protocol TLWebViewDelegate: class {
    func webView(webView: TLWebView, didProposeVisitToLocation location: NSURL)
    func webViewDidInvalidatePage(webView: TLWebView)
}

protocol TLWebViewVisitDelegate: class {
    func webView(webView: TLWebView, didStartVisitWithIdentifier identifier: String, hasSnapshot: Bool)
    func webView(webView: TLWebView, didRestoreSnapshotForVisitWithIdentifier identifier: String)
    func webView(webView: TLWebView, didStartRequestForVisitWithIdentifier identifier: String)
    func webView(webView: TLWebView, didCompleteRequestForVisitWithIdentifier identifier: String)
    func webView(webView: TLWebView, didFailRequestForVisitWithIdentifier identifier: String, withStatusCode statusCode: Int?)
    func webView(webView: TLWebView, didFinishRequestForVisitWithIdentifier identifier: String)
    func webView(webView: TLWebView, didLoadResponseForVisitWithIdentifier identifier: String)
    func webView(webView: TLWebView, didCompleteVisitWithIdentifier identifier: String)
}

class TLWebView: WKWebView, WKScriptMessageHandler {
    weak var delegate: TLWebViewDelegate?
    weak var visitDelegate: TLWebViewVisitDelegate?

    init(configuration: WKWebViewConfiguration) {
        super.init(frame: CGRectZero, configuration: configuration)

        let bundle = NSBundle(forClass: self.dynamicType)
        let source = try! String(contentsOfURL: bundle.URLForResource("TLWebView", withExtension: "js")!, encoding: NSUTF8StringEncoding)
        let userScript = WKUserScript(source: source, injectionTime: .AtDocumentEnd, forMainFrameOnly: true)
        configuration.userContentController.addUserScript(userScript)
        configuration.userContentController.addScriptMessageHandler(self, name: "turbolinks")

        translatesAutoresizingMaskIntoConstraints = false
        scrollView.decelerationRate = UIScrollViewDecelerationRateNormal
    }

    func visitLocation(location: NSURL, withAction action: String) {
        callJavaScriptFunction("webView.visitLocationWithAction", withArguments: [location.absoluteString, action])
    }

    func issueRequestForVisitWithIdentifier(identifier: String) {
        callJavaScriptFunction("webView.issueRequestForVisitWithIdentifier", withArguments: [identifier])
    }

    func changeHistoryForVisitWithIdentifier(identifier: String) {
        callJavaScriptFunction("webView.changeHistoryForVisitWithIdentifier", withArguments: [identifier])
    }

    func restoreSnapshotForVisitWithIdentifier(identifier: String) {
        callJavaScriptFunction("webView.restoreSnapshotForVisitWithIdentifier", withArguments: [identifier])
    }

    func loadResponseForVisitWithIdentifier(identifier: String) {
        callJavaScriptFunction("webView.loadResponseForVisitWithIdentifier", withArguments: [identifier])
    }

    func cancelVisitWithIdentifier(identifier: String) {
        callJavaScriptFunction("webView.cancelVisitWithIdentifier", withArguments: [identifier])
    }

    // MARK: WKScriptMessageHandler

    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
        if let message = TLScriptMessage.parse(message) {
            switch message.name {
            case .VisitProposed:
                delegate?.webView(self, didProposeVisitToLocation: message.location!)
            case .PageInvalidated:
                delegate?.webViewDidInvalidatePage(self)
            case .VisitStarted:
                visitDelegate?.webView(self, didStartVisitWithIdentifier: message.identifier!, hasSnapshot: message.data["hasSnapshot"] as! Bool)
            case .VisitSnapshotRestored:
                visitDelegate?.webView(self, didRestoreSnapshotForVisitWithIdentifier: message.identifier!)
            case .VisitRequestStarted:
                visitDelegate?.webView(self, didStartRequestForVisitWithIdentifier: message.identifier!)
            case .VisitRequestCompleted:
                visitDelegate?.webView(self, didCompleteRequestForVisitWithIdentifier: message.identifier!)
            case .VisitRequestFailed:
                let statusCode = message.data["statusCode"] as? Int
                visitDelegate?.webView(self, didFailRequestForVisitWithIdentifier: message.identifier!, withStatusCode: statusCode)
            case .VisitRequestFinished:
                visitDelegate?.webView(self, didFinishRequestForVisitWithIdentifier: message.identifier!)
            case .VisitResponseLoaded:
                visitDelegate?.webView(self, didLoadResponseForVisitWithIdentifier: message.identifier!)
            case .VisitCompleted:
                visitDelegate?.webView(self, didCompleteVisitWithIdentifier: message.identifier!)
            case .Error:
                let error = message.data["error"] as? String
                NSLog("JavaScript error: \(error)")
            }
        }
    }

    // MARK: JavaScript Evaluation

    private func callJavaScriptFunction(functionExpression: String, withArguments arguments: [AnyObject] = [], completionHandler: ((AnyObject?) -> ())? = nil) {
        if let script = scriptForCallingJavaScriptFunction(functionExpression, withArguments: arguments) {
            evaluateJavaScript(script) { (result, error) in
                if let result = result as? Dictionary<String, AnyObject> {
                    if let error = result["error"] as? String, stack = result["stack"] as? String {
                        NSLog("Error evaluating JavaScript function `\(functionExpression)': \(error)\n\(stack)")
                    } else {
                        completionHandler?(result["value"])
                    }
                }

            }
        } else {
            NSLog("Error encoding arguments for JavaScript function `\(functionExpression)'")
        }
    }

    private func scriptForCallingJavaScriptFunction(functionExpression: String, withArguments arguments: [AnyObject]) -> String? {
        if let encodedArguments = encodeJavaScriptArguments(arguments) {
            return
                "(function(result) {\n" +
                "  try {\n" +
                "    result.value = " + functionExpression + "(" + encodedArguments + ")\n" +
                "  } catch (error) {\n" +
                "    result.error = error.toString()\n" +
                "    result.stack = error.stack\n" +
                "  }\n" +
                "  return result\n" +
                "})({})"
        }
        return nil
    }

    private func encodeJavaScriptArguments(arguments: [AnyObject]) -> String? {
        if let data = try? NSJSONSerialization.dataWithJSONObject(arguments, options: []),
            string = NSString(data: data, encoding: NSUTF8StringEncoding) as? String {
                return string[Range(start: string.startIndex.successor(), end: string.endIndex.predecessor())]
        }
        return nil
    }
}

File: Turbolinks/Turbolinks/TLWebView.js
function TLWebView(controller, messageHandler) {
    this.controller = controller
    this.messageHandler = messageHandler
    controller.adapter = this
}

TLWebView.prototype = {
    visitLocationWithAction: function(location, action) {
        this.controller.startVisitToLocationWithAction(location, action)
    },

    // Current visit

    issueRequestForVisitWithIdentifier: function(identifier) {
        if (identifier == this.currentVisit.identifier) {
            this.currentVisit.issueRequest()
        }
    },

    changeHistoryForVisitWithIdentifier: function(identifier) {
        if (identifier == this.currentVisit.identifier) {
            this.currentVisit.changeHistory()
        }
    },

    restoreSnapshotForVisitWithIdentifier: function(identifier) {
        if (identifier == this.currentVisit.identifier) {
            this.currentVisit.restoreSnapshot()
        }
    },

    loadResponseForVisitWithIdentifier: function(identifier) {
        if (identifier == this.currentVisit.identifier) {
            this.currentVisit.loadResponse()
        }
    },

    cancelVisitWithIdentifier: function(identifier) {
        if (identifier == this.currentVisit.identifier) {
            this.currentVisit.cancel()
        }
    },

    // Adapter interface
   
    visitProposedToLocationWithAction: function(location, action) {
        this.postMessage("visitProposed", { location: location.absoluteURL, action: action })
    },

    visitStarted: function(visit) {
        this.currentVisit = visit
        this.postMessage("visitStarted", { identifier: visit.identifier, hasSnapshot: visit.hasSnapshot() })
    },

    visitRequestStarted: function(visit) {
        this.postMessage("visitRequestStarted", { identifier: visit.identifier })
    },

    visitRequestCompleted: function(visit) {
        this.postMessage("visitRequestCompleted", { identifier: visit.identifier })
    },

    visitRequestFailedWithStatusCode: function(visit, statusCode) {
        this.postMessage("visitRequestFailed", { identifier: visit.identifier, statusCode: statusCode })
    },

    visitRequestFinished: function(visit) {
        this.postMessage("visitRequestFinished", { identifier: visit.identifier })
    },

    visitSnapshotRestored: function(visit) {
        this.postMessageAfterNextRepaint("visitSnapshotRestored", { identifier: visit.identifier })
    },

    visitResponseLoaded: function(visit) {
        this.postMessageAfterNextRepaint("visitResponseLoaded", { identifier: visit.identifier })
    },

    visitCompleted: function(visit) {
        this.postMessageAfterNextRepaint("visitCompleted", { identifier: visit.identifier })
    },

    pageInvalidated: function() {
        this.postMessage("pageInvalidated")
    },

    // Private

    postMessage: function(name, data) {
        this.messageHandler.postMessage({ name: name, data: data || {} })
    },

    postMessageAfterNextRepaint: function(name, data) {
        var postMessage = this.postMessage.bind(this, name, data)
        requestAnimationFrame(function() {
            requestAnimationFrame(postMessage)
        })
    }
}

window.webView = new TLWebView(Turbolinks.controller, webkit.messageHandlers.turbolinks)

addEventListener("error", function(event) {
    var error = event.message + " (" + event.filename + ":" + event.lineno + ":" + event.colno + ")"
    webView.postMessage("error", { error: error })
}, false)
